### ISSUE TITLE
Put the path choices in clients

### ISSUE THREAD

**Issue by cwgoes** [cosmos/ibc#288]
Instead of forcing specific paths to be checked.

This would make the solo-chain case work better, it's more flexible.

Thought after a conversation with Dean from Agoric.

**Comment by cwgoes** [comment]
Then the current layout would turn into a suggested client pattern, just like general Merkle proofs.

**Comment by cwgoes** [comment]
This will mean that `verifyMembership` / `verifyNonMembership` on clients will be replaced with:
- `verifyChannelState`
- `verifyConnectionState`
- `verifyClientConsensusState`
- `verifyOutgoingPacket`
- `verifyPacketAcknowledgement`
- `verifyNextSequenceRecv`

They will still take a height, an opaque proof, and various path parameters. Instead of interpolating the path string, however, clients will be able to implement the functions any way they want (including in ways that don't require a Merkle-ized state, such as just checking a signature in `verifyOutgoingPacket`).

The current store layout can still be a default, recommended for clients with Merkle-ized state.

**Comment by fedekunze** [comment]
@cwgoes are there any cons of doing this besides replacing the generic verification functions?

**Comment by cwgoes** [comment]
I don't believe so, it doesn't change the exposed interface at all, just where the store keys are defined.

**Comment by cwgoes** [comment]
Should connections still be able to provide a parameter to these functions?

**Comment by cwgoes** [comment]
I think the answer is yes. Are there examples of what it would be other than a sub-path?

**Comment by fedekunze** [comment]
What do you mean with providing a parameter?

**Comment by cwgoes** [comment]
Right now, connections also have `verifyMembership` / `verifyNonMembership`, which just calls the associated client object, passing a `Path []byte` corresponding to the connection. This allows the same client to be used with multiple connections with different store paths (e.g. multiple IBC modules on the same chain), which could be useful.

**Comment by gjermundgaraba** [comment]
This issue was mistakenly reopened during some integration work, so closing again now.
